R/verb-select.R

```diff
@@ -104,37 +104,66 @@ simulate_vars <- function(x, drop_groups = FALSE) {
   as_tibble(rep_named(vars, list(logical())), .name_repair = "minimal")
 }
 
+simulate_vars2 <- function(x) {
+  vars <- op_vars(x)
+  vars_list <- as.list(vars)
+  as_tibble(set_names(vars_list, vars))
+}
+
 # op_select ---------------------------------------------------------------
 
-op_select <- function(x, vars) {
-  if (selects_same_variables(x, vars)) {
-    return(x)
+add_select <- function(.data, vars, op = c("select", "mutate")) {
+  op <- match.arg(op, c("select", "mutate"))
+  lazy_query <- .data$lazy_query
+
+  # drop NULLs
+  vars <- purrr::discard(vars, ~ is_quosure(.x) && quo_is_null(.x))
+  if (selects_same_variables(.data, vars)) {
+    return(lazy_query)
   }
 
-  if (inherits(x, "op_select")) {
+  if (identical(lazy_query$last_op, "select") || identical(lazy_query$last_op, "mutate")) {
```

Comment:
`lazy_query$last_op %in% c("select", "mutate")`, since `last_op` is always a string?

Reply:
The `last_op` field only exists for `lazy_select_query`, so `lazy_query$last_op` can also be `NULL`.